File: Tutorials/Codes/System_Dynamics_Course/Chapter14/Lesson5/Chapter14_Lesson5.py
# Chapter14_Lesson5.py
# Limit Cycles, Multiple Equilibria, and Basic Bifurcation Notions
# Python implementation: Van der Pol oscillator + 1D bifurcation normal forms
import numpy as np
import matplotlib.pyplot as plt

def rk4_step(f, t, x, h, params):
    k1 = f(t, x, params)
    k2 = f(t + 0.5*h, x + 0.5*h*k1, params)
    k3 = f(t + 0.5*h, x + 0.5*h*k2, params)
    k4 = f(t + h, x + h*k3, params)
    return x + (h/6.0)*(k1 + 2*k2 + 2*k3 + k4)

def vdp_rhs(t, state, params):
    mu = params["mu"]
    x, y = state
    dx = y
    dy = mu*(1.0 - x*x)*y - x
    return np.array([dx, dy], dtype=float)

def simulate_vdp(mu=1.0, x0=(2.0, 0.0), T=80.0, h=0.01):
    n = int(T/h)
    t = np.linspace(0.0, T, n+1)
    x = np.zeros((n+1, 2))
    x[0] = np.array(x0, dtype=float)
    params = {"mu": mu}
    for k in range(n):
        x[k+1] = rk4_step(vdp_rhs, t[k], x[k], h, params)
    return t, x

def estimate_period_from_crossings(t, traj):
    # Upward crossings of y=0 (x-axis): use y sign change from negative to positive
    y = traj[:, 1]
    x = traj[:, 0]
    crossings = []
    for k in range(len(t)-1):
        if y[k] < 0.0 and y[k+1] >= 0.0:
            # Linear interpolation
            alpha = -y[k] / (y[k+1] - y[k] + 1e-14)
            tc = t[k] + alpha*(t[k+1] - t[k])
            xc = x[k] + alpha*(x[k+1] - x[k])
            if xc > 0:  # consistent Poincare section branch
                crossings.append(tc)
    if len(crossings) < 3:
        return np.nan
    # Discard transient crossing intervals
    periods = np.diff(crossings)
    if len(periods) >= 4:
        periods = periods[-4:]
    return float(np.mean(periods))

def classify_pitchfork(r, x):
    # f(x)=r x - x^3 ; stability from f'(x)=r - 3 x^2
    lam = r - 3.0*x*x
    if lam < 0:
        return "stable"
    elif lam > 0:
        return "unstable"
    return "nonhyperbolic"

def classify_saddlenode(r, x):
    # f(x)=r - x^2 ; stability from f'(x)=-2x
    lam = -2.0*x
    if lam < 0:
        return "stable"
    elif lam > 0:
        return "unstable"
    return "nonhyperbolic"

def main():
    # 1) Van der Pol simulation for several mu values
    mu_list = [0.2, 1.0, 3.0]
    results = []
    fig1 = plt.figure(figsize=(7, 5))
    for mu in mu_list:
        t, z = simulate_vdp(mu=mu, x0=(2.0, 0.1), T=80.0, h=0.01)
        # discard transient for plotting
        z_ss = z[int(0.5*len(z)):, :]
        plt.plot(z_ss[:,0], z_ss[:,1], label=f"mu={mu}")
        period = estimate_period_from_crossings(t[int(0.4*len(t)):], z[int(0.4*len(z)):, :])
        amp = float(np.max(np.abs(z_ss[:,0])))
        results.append((mu, amp, period))
    plt.xlabel("x")
    plt.ylabel("x_dot")
    plt.title("Van der Pol Limit Cycles (phase plane)")
    plt.legend()
    plt.grid(True)
    plt.tight_layout()

    # 2) 1D bifurcation normal forms (equilibrium branches)
    r_vals = np.linspace(-2.0, 2.0, 401)
    pf_stable_r, pf_stable_x = [], []
    pf_unstable_r, pf_unstable_x = [], []
    sn_stable_r, sn_stable_x = [], []
    sn_unstable_r, sn_unstable_x = [], []

    # Pitchfork: x*=0 and ±sqrt(r) for r>=0
    for r in r_vals:
        for xeq in [0.0]:
            c = classify_pitchfork(r, xeq)
            (pf_stable_r if c == "stable" else pf_unstable_r).append(r)
            (pf_stable_x if c == "stable" else pf_unstable_x).append(xeq)
        if r >= 0:
            for xeq in [np.sqrt(r), -np.sqrt(r)]:
                c = classify_pitchfork(r, xeq)
                (pf_stable_r if c == "stable" else pf_unstable_r).append(r)
                (pf_stable_x if c == "stable" else pf_unstable_x).append(xeq)

    # Saddle-node: x*=±sqrt(r) for r>=0
    for r in r_vals:
        if r >= 0:
            for xeq in [np.sqrt(r), -np.sqrt(r)]:
                c = classify_saddlenode(r, xeq)
                (sn_stable_r if c == "stable" else sn_unstable_r).append(r)
                (sn_stable_x if c == "stable" else sn_unstable_x).append(xeq)

    fig2 = plt.figure(figsize=(7, 5))
    plt.plot(pf_stable_r, pf_stable_x, linewidth=2, label="Pitchfork stable")
    plt.plot(pf_unstable_r, pf_unstable_x, "--", linewidth=2, label="Pitchfork unstable")
    plt.plot(sn_stable_r, sn_stable_x, linewidth=2, label="Saddle-node stable")
    plt.plot(sn_unstable_r, sn_unstable_x, "--", linewidth=2, label="Saddle-node unstable")
    plt.xlabel("parameter r")
    plt.ylabel("equilibrium x*")
    plt.title("Basic Bifurcation Branches")
    plt.legend()
    plt.grid(True)
    plt.tight_layout()

    # 3) Print numerical summary
    print("Van der Pol asymptotic limit-cycle estimates")
    print("mu\tamplitude(|x|)\tperiod")
    for mu, amp, period in results:
        print(f"{mu:.2f}\t{amp:.4f}\t\t{period:.4f}")

    plt.show()

if __name__ == "__main__":
    main()

File: Tutorials/Codes/System_Dynamics_Course/Chapter14/Lesson5/Chapter14_Lesson5.m
% Chapter14_Lesson5.m
% Limit Cycles, Multiple Equilibria, and Basic Bifurcation Notions
% MATLAB + basic Simulink automation (if Simulink is installed)

clear; clc; close all;

%% Part A: Van der Pol oscillator (ode45) and limit-cycle period estimation
mu = 1.0;
f = @(t,z) [z(2); mu*(1 - z(1)^2)*z(2) - z(1)];
tspan = [0 80];
z0 = [2; 0.1];
opts = odeset('RelTol',1e-8,'AbsTol',1e-10);
[t,z] = ode45(f, tspan, z0, opts);

% phase portrait (transient removed)
idx0 = floor(numel(t)/2);
figure('Name','Van der Pol Phase Portrait');
plot(z(idx0:end,1), z(idx0:end,2), 'LineWidth', 1.5); grid on;
xlabel('x'); ylabel('x_dot');
title('Van der Pol Limit Cycle (mu = 1)');

% Poincare crossing estimate: y = 0 upward, x > 0
crossings = [];
for k = idx0:(numel(t)-1)
    if z(k,2) < 0 && z(k+1,2) >= 0
        alpha = -z(k,2)/(z(k+1,2)-z(k,2) + eps);
        tc = t(k) + alpha*(t(k+1)-t(k));
        xc = z(k,1) + alpha*(z(k+1,1)-z(k,1));
        if xc > 0
            crossings(end+1) = tc; %#ok<SAGROW>
        end
    end
end
if numel(crossings) >= 3
    P = mean(diff(crossings(max(1,end-4):end)));
else
    P = NaN;
end
A = max(abs(z(idx0:end,1)));
fprintf('Van der Pol limit-cycle estimate: amplitude = %.6f, period = %.6f\n', A, P);

%% Part B: Basic 1D bifurcation normal forms
r = linspace(-2, 2, 801);
pf_stable = []; pf_unstable = [];
sn_stable = []; sn_unstable = [];

for rv = r
    % Pitchfork: x*=0 always
    lam0 = rv;
    if lam0 < 0
        pf_stable = [pf_stable; rv 0]; %#ok<AGROW>
    else
        pf_unstable = [pf_unstable; rv 0]; %#ok<AGROW>
    end
    if rv >= 0
        xe = [sqrt(rv), -sqrt(rv)];
        for xeq = xe
            lam = rv - 3*xeq^2;
            if lam < 0
                pf_stable = [pf_stable; rv xeq]; %#ok<AGROW>
            else
                pf_unstable = [pf_unstable; rv xeq]; %#ok<AGROW>
            end
        end
    end

    % Saddle-node: x*= +-sqrt(r) for r>=0
    if rv >= 0
        xplus = sqrt(rv); xminus = -sqrt(rv);
        if -2*xplus < 0, sn_stable = [sn_stable; rv xplus]; else, sn_unstable = [sn_unstable; rv xplus]; end %#ok<AGROW>
        if -2*xminus < 0, sn_stable = [sn_stable; rv xminus]; else, sn_unstable = [sn_unstable; rv xminus]; end %#ok<AGROW>
    end
end

figure('Name','Bifurcation Branches');
hold on; grid on;
if ~isempty(pf_stable),   plot(pf_stable(:,1), pf_stable(:,2),  'LineWidth', 1.6); end
if ~isempty(pf_unstable), plot(pf_unstable(:,1), pf_unstable(:,2),'--','LineWidth', 1.2); end
if ~isempty(sn_stable),   plot(sn_stable(:,1), sn_stable(:,2),   'LineWidth', 1.6); end
if ~isempty(sn_unstable), plot(sn_unstable(:,1), sn_unstable(:,2),'--','LineWidth', 1.2); end
xlabel('r'); ylabel('x^*');
title('Pitchfork and Saddle-Node Equilibrium Branches');
legend('Pitchfork stable','Pitchfork unstable','Saddle-node stable','Saddle-node unstable');

%% Part C: Simulink (optional) - programmatically build a Van der Pol model
% This section creates a simple Simulink model with two integrators if Simulink exists.
if license('test','Simulink') && exist('new_system','file')
    mdl = 'Chapter14_Lesson5_SimulinkModel';
    if bdIsLoaded(mdl), close_system(mdl,0); end
    if exist([mdl '.slx'],'file'), delete([mdl '.slx']); end
    new_system(mdl); open_system(mdl);

    add_block('simulink/Sources/Constant', [mdl '/mu'], 'Position', [30 40 60 60], 'Value', '1');
    add_block('simulink/Continuous/Integrator', [mdl '/Int_x'], 'Position', [420 90 450 120]);
    add_block('simulink/Continuous/Integrator', [mdl '/Int_y'], 'Position', [420 180 450 210]);

    add_block('simulink/Math Operations/Product', [mdl '/x2'], 'Position', [120 140 150 170], 'Inputs', '**');
    add_block('simulink/Math Operations/Sum', [mdl '/1_minus_x2'], 'Position', [180 140 205 170], 'Inputs', '+-');
    add_block('simulink/Sources/Constant', [mdl '/one'], 'Position', [120 90 150 110], 'Value', '1');

    add_block('simulink/Math Operations/Product', [mdl '/mu_term'], 'Position', [240 120 270 150], 'Inputs', '**');
    add_block('simulink/Math Operations/Product', [mdl '/mu_term_times_y'], 'Position', [300 120 330 150], 'Inputs', '**');
    add_block('simulink/Math Operations/Gain', [mdl '/minus_x'], 'Position', [300 200 340 230], 'Gain', '-1');
    add_block('simulink/Math Operations/Sum', [mdl '/ydot_sum'], 'Position', [360 155 385 185], 'Inputs', '++');

    add_block('simulink/Sinks/Scope', [mdl '/Scope'], 'Position', [520 90 550 120]);

    % Connect x and y states
    add_line(mdl, 'Int_x/1', 'x2/1');
    add_line(mdl, 'Int_x/1', 'x2/2');
    add_line(mdl, 'one/1', '1_minus_x2/1');
    add_line(mdl, 'x2/1', '1_minus_x2/2');
    add_line(mdl, 'mu/1', 'mu_term/1');
    add_line(mdl, '1_minus_x2/1', 'mu_term/2');
    add_line(mdl, 'mu_term/1', 'mu_term_times_y/1');
    add_line(mdl, 'Int_y/1', 'mu_term_times_y/2');
    add_line(mdl, 'mu_term_times_y/1', 'ydot_sum/1');
    add_line(mdl, 'Int_x/1', 'minus_x/1');
    add_line(mdl, 'minus_x/1', 'ydot_sum/2');

    % State equations: xdot = y, ydot = mu(1-x^2)y - x
    add_line(mdl, 'Int_y/1', 'Int_x/1', 'autorouting', 'on');
    add_line(mdl, 'ydot_sum/1', 'Int_y/1', 'autorouting', 'on');
    add_line(mdl, 'Int_x/1', 'Scope/1', 'autorouting', 'on');

    set_param(mdl, 'StopTime', '40');
    save_system(mdl);
    fprintf('Simulink model created: %s.slx\n', mdl);
else
    fprintf('Simulink not available. MATLAB ODE implementation executed.\n');
end
